### ISSUE TITLE
Broadcast ops with Arrays

### ISSUE THREAD

**Issue by dinarior** [JuliaGPU/CuArrays.jl#241]
While:

```
arr = ones(10)
arr2 = ones(1)
CuArray(arr) .- CuArray(arr2)
```
 Works fine, `CuArray(arr) .- arr2` fails. 
Is that intended?

**Comment by vchuravy** [comment]
Yes. It is unclear (at least to me) how to handle heterogenous broadcast/operations. Would we copy the GPU array to the host or the host array to the GPU? What if there are 9 CPU arrays and 1 GPU Arrays?

**Comment by MikeInnes** [comment]
Ideally I guess you'd go by size or something but it'd no longer be type inferable then.

I think just deciding on "always move to GPU" here would be perfectly reasonable. A majority vote of all internal arrays would actually be doable as well.

**Comment by ali-ramadhan** [comment]
Would it make sense to provide this for expressions such as `array1 .= array2`? Not sure if only specific/simple broadcast operations can be supported.

In this case it's a bit more obvious. If `array1` is a `CuArray` and `array2` is an `Array` then the contents of `array2` are copied from the CPU to the GPU, and vice versa.

**Comment by MikeInnes** [comment]
Yes, in in-place cases you'd have to move to wherever the target array is. That might not be too hard to implement.

**Comment by vchuravy** [comment]
If you want that I would recommend explicitly using `copyto!` instead of implicitly broadcasting `identity`
